File: 357-count-numbers-with-unique-digits/leochang/357.cpp
class Solution {
public:
	int countNumbersWithUniqueDigits(int n) {
		if (n == 0) return 1;
		int ans = 1, p = 9, unique = 9;
		int m = min(n, 10);
		for (int i = 0; i < n; i++) {
			ans += unique;
			unique *= p;
			p--;
		}

		return ans;
	}
};

File: 357-count-numbers-with-unique-digits/ken/357.py
class Solution(object):
    def countNumbersWithUniqueDigits(self, n):
        """
        :type n: int
        :rtype: int
        """

        def countExactDigits(numOfDigits):
            if numOfDigits == 1:
                return 10

            product = 1
            numOfCandidate = [9] + [9 - i for i in xrange(9)]
            for i in xrange(numOfDigits):
                product *= numOfCandidate[i]

            return product

        if n == 0:
            return 1

        ans = 0
        for i in xrange(n):
            ans += countExactDigits(i + 1)

        return ans
